<p style="text-align:center">
    <a href="https://skills.network" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>

<h1 align=center><font size = 5>Assignment: SQL Notebook for Peer Assignment</font></h1>

Estimated time needed: **60** minutes.

## Introduction
Using this Python notebook you will:

1.  Understand the Spacex DataSet
2.  Load the dataset  into the corresponding table in a Db2 database
3.  Execute SQL queries to answer assignment questions


## Overview of the DataSet

SpaceX has gained worldwide attention for a series of historic milestones.

It is the only private company ever to return a spacecraft from low-earth orbit, which it first accomplished in December 2010.
SpaceX advertises Falcon 9 rocket launches on its website with a cost of 62 million dollars wheras other providers cost upward of 165 million dollars each, much of the savings is because Space X can reuse the first stage.


Therefore if we can determine if the first stage will land, we can determine the cost of a launch.

This information can be used if an alternate company wants to bid against SpaceX for a rocket launch.

This dataset includes a record for each payload carried during a SpaceX mission into outer space.


### Download the datasets

This assignment requires you to load the spacex dataset.

In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. Click on the link below to download and save the dataset (.CSV file):

 <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_2/data/Spacex.csv" target="_blank">Spacex DataSet</a>



In [1]:
!pip install sqlalchemy==1.3.9


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 11.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for sqlalchemy: filename=SQLAlchemy-1.3.9-cp310-cp310-linux_x86_64.whl size=1192330 sha256=0b12160b5e9eea07444e9be6a82efde94420373d4a863b5cf817325f7febe18b
  Stored in directory: /root/.cache/pip/wheels/8b/6d/9c/e1793e5a2d85b89b8c362a4f0e3def54b786bcd14923b9ff00
Successfully built sqlalchemy
  Attempting uninstall: sqlalchemy
    Found existing installation: SQLAlchemy 2.0.31
    Uninstalling SQLAlchemy-2.0.31:
      Successfully uninstalled SQLAlchemy-2.0.31
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
bigframes 1.11.1 requires sqlalchemy<3.0dev,>=1.4, but you have sqlalchemy 1.3.9 which is incompatible.
ipython-sql 0.5.0 requires sqlalchemy>=2.0, but you have sqlalchemy 1.3.9 which is incompatible.


### Connect to the database

Let us first load the SQL extension and establish a connection with the database


In [2]:
!pip install ipython-sql

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 11.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 24.7 MB/s eta 0:00:00
  Attempting uninstall: sqlalchemy
    Found existing installation: SQLAlchemy 1.3.9
    Uninstalling SQLAlchemy-1.3.9:
      Successfully uninstalled SQLAlchemy-1.3.9


In [3]:
%load_ext sql

In [4]:
import csv, sqlite3

con = sqlite3.connect("my_data1.db")
cur = con.cursor()

In [5]:
!pip install -q pandas

In [6]:
%sql sqlite:///my_data1.db

In [8]:
import pandas as pd
df = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_2/data/Spacex.csv")
df.to_sql("SPACEXTBL", con, if_exists='replace', index=False,method="multi")

101

**Note:This below code is added to remove blank rows from table**


In [9]:
df.head()

Date Time (UTC) Booster_Version  Launch_Site  \
0  2010-06-04   18:45:00  F9 v1.0  B0003  CCAFS LC-40   
1  2010-12-08   15:43:00  F9 v1.0  B0004  CCAFS LC-40   
2  2012-05-22    7:44:00  F9 v1.0  B0005  CCAFS LC-40   
3  2012-10-08    0:35:00  F9 v1.0  B0006  CCAFS LC-40   
4  2013-03-01   15:10:00  F9 v1.0  B0007  CCAFS LC-40   

                                             Payload  PAYLOAD_MASS__KG_  \
0               Dragon Spacecraft Qualification Unit                  0   
1  Dragon demo flight C1, two CubeSats, barrel of...                  0   
2                              Dragon demo flight C2                525   
3                                       SpaceX CRS-1                500   
4                                       SpaceX CRS-2                677   

       Orbit         Customer Mission_Outcome      Landing_Outcome  
0        LEO           SpaceX         Success  Failure (parachute)  
1  LEO (ISS)  NASA (COTS) NRO         Success  Failure (parachute)  
2  LEO (ISS)      NASA (COTS)         Success           No attempt  
3  LEO (ISS)       NASA (CRS)         Success           No attempt  
4  LEO (ISS)       NASA (CRS)         Success           No attempt

In [10]:
#DROP THE TABLE IF EXISTS

%sql DROP TABLE IF EXISTS SPACEXTABLE;

 * sqlite:///my_data1.db
Done.


[]

In [11]:
%sql create table SPACEXTABLE as select * from SPACEXTBL where Date is not null

 * sqlite:///my_data1.db
Done.


[]

## Tasks

Now write and execute SQL queries to solve the assignment tasks.

**Note: If the column names are in mixed case enclose it in double quotes
   For Example "Landing_Outcome"**

### Task 1




##### Display the names of the unique launch sites  in the space mission


In [13]:
%sql select distinct Launch_Site from SPACEXTABLE;

 * sqlite:///my_data1.db
Done.


[('CCAFS LC-40',), ('VAFB SLC-4E',), ('KSC LC-39A',), ('CCAFS SLC-40',)]


### Task 2


#####  Display 5 records where launch sites begin with the string 'CCA'


In [14]:
%sql select * from SPACEXTABLE where Launch_Site like 'CCA%' limit 5;

 * sqlite:///my_data1.db
Done.


[('2010-06-04', '18:45:00', 'F9 v1.0  B0003', 'CCAFS LC-40', 'Dragon Spacecraft Qualification Unit', 0, 'LEO', 'SpaceX', 'Success', 'Failure (parachute)'),
 ('2010-12-08', '15:43:00', 'F9 v1.0  B0004', 'CCAFS LC-40', 'Dragon demo flight C1, two CubeSats, barrel of Brouere cheese', 0, 'LEO (ISS)', 'NASA (COTS) NRO', 'Success', 'Failure (parachute)'),
 ('2012-05-22', '7:44:00', 'F9 v1.0  B0005', 'CCAFS LC-40', 'Dragon demo flight C2', 525, 'LEO (ISS)', 'NASA (COTS)', 'Success', 'No attempt'),
 ('2012-10-08', '0:35:00', 'F9 v1.0  B0006', 'CCAFS LC-40', 'SpaceX CRS-1', 500, 'LEO (ISS)', 'NASA (CRS)', 'Success', 'No attempt'),
 ('2013-03-01', '15:10:00', 'F9 v1.0  B0007', 'CCAFS LC-40', 'SpaceX CRS-2', 677, 'LEO (ISS)', 'NASA (CRS)', 'Success', 'No attempt')]

### Task 3




##### Display the total payload mass carried by boosters launched by NASA (CRS)


In [20]:
%sql select sum("PAYLOAD_MASS__KG_") from SPACEXTABLE where Customer = 'NASA (CRS)';

 * sqlite:///my_data1.db
Done.


[(45596,)]

### Task 4




##### Display average payload mass carried by booster version F9 v1.1


In [23]:
%sql select avg("PAYLOAD_MASS_KG_") from SPACEXTABLE where Booster_Version = 'F9 v1.1';

 * sqlite:///my_data1.db
Done.


[(0.0,)]

In [25]:
%sql select * from SPACEXTABLE where "Booster_Version" = 'F9 v1.1' and "PAYLOAD_MASS_KG_">0;

 * sqlite:///my_data1.db
Done.


[('2013-12-03', '22:41:00', 'F9 v1.1', 'CCAFS LC-40', 'SES-8', 3170, 'GTO', 'SES', 'Success', 'No attempt'),
 ('2014-01-06', '22:06:00', 'F9 v1.1', 'CCAFS LC-40', 'Thaicom 6', 3325, 'GTO', 'Thaicom', 'Success', 'No attempt'),
 ('2014-04-18', '19:25:00', 'F9 v1.1', 'CCAFS LC-40', 'SpaceX CRS-3', 2296, 'LEO (ISS)', 'NASA (CRS)', 'Success', 'Controlled (ocean)'),
 ('2014-07-14', '15:15:00', 'F9 v1.1', 'CCAFS LC-40', 'OG2 Mission 1  6 Orbcomm-OG2 satellites', 1316, 'LEO', 'Orbcomm', 'Success', 'Controlled (ocean)'),
 ('2014-08-05', '8:00:00', 'F9 v1.1', 'CCAFS LC-40', 'AsiaSat 8', 4535, 'GTO', 'AsiaSat', 'Success', 'No attempt')]

### Task 5

##### List the date when the first succesful landing outcome in ground pad was acheived.


_Hint:Use min function_


In [26]:
%sql select min("Date") from SPACEXTABLE where "Landing_Outcome" = 'Success (ground pad)';

 * sqlite:///my_data1.db
Done.


[('2015-12-22',)]

### Task 6

##### List the names of the boosters which have success in drone ship and have payload mass greater than 4000 but less than 6000


In [30]:
%sql select "Booster_Version" from SPACEXTABLE where "Landing_Outcome" = 'Success (drone ship)' and "PAYLOAD_MASS_KG_" between 4000 and 6000;

 * sqlite:///my_data1.db
Done.


[]

### Task 7




##### List the total number of successful and failure mission outcomes


In [31]:
%sql select "Mission_Outcome", count(*) from SPACEXTABLE group by "Mission_Outcome";

 * sqlite:///my_data1.db
Done.


[('Failure (in flight)', 1),
 ('Success', 98),
 ('Success ', 1),
 ('Success (payload status unclear)', 1)]

### Task 8



##### List the   names of the booster_versions which have carried the maximum payload mass. Use a subquery


In [34]:
%sql select "Booster_Version" from SPACEXTABLE where "PAYLOAD_MASS_KG_"= (select max("PAYLOAD_MASS_KG_") from SPACEXTABLE)

 * sqlite:///my_data1.db
Done.


[('F9 v1.0  B0003',),
 ('F9 v1.0  B0004',),
 ('F9 v1.0  B0005',),
 ('F9 v1.0  B0006',),
 ('F9 v1.0  B0007',),
 ('F9 v1.1  B1003',),
 ('F9 v1.1',),
 ('F9 v1.1',),
 ('F9 v1.1',),
 ('F9 v1.1',),
 ('F9 v1.1',),
 ('F9 v1.1 B1011',),
 ('F9 v1.1 B1010',),
 ('F9 v1.1 B1012',),
 ('F9 v1.1 B1013',),
 ('F9 v1.1 B1014',),
 ('F9 v1.1 B1015',),
 ('F9 v1.1 B1016',),
 ('F9 v1.1 B1018',),
 ('F9 FT B1019',),
 ('F9 v1.1 B1017',),
 ('F9 FT B1020',),
 ('F9 FT B1021.1',),
 ('F9 FT B1022',),
 ('F9 FT B1023.1',),
 ('F9 FT B1024',),
 ('F9 FT B1025.1',),
 ('F9 FT B1026',),
 ('F9 FT B1029.1',),
 ('F9 FT B1031.1',),
 ('F9 FT B1030',),
 ('F9 FT  B1021.2',),
 ('F9 FT B1032.1',),
 ('F9 FT B1034',),
 ('F9 FT B1035.1',),
 ('F9 FT  B1029.2',),
 ('F9 FT B1036.1',),
 ('F9 FT B1037',),
 ('F9 B4 B1039.1',),
 ('F9 FT B1038.1',),
 ('F9 B4 B1040.1',),
 ('F9 B4 B1041.1',),
 ('F9 FT  B1031.2',),
 ('F9 B4 B1042.1',),
 ('F9 FT  B1035.2',),
 ('F9 FT  B1036.2',),
 ('F9 B4 B1043.1',),
 ('F9 FT  B1032.2',),
 ('F9 FT  B1038.2',),
 ('F9 B4 B1044',),
 ('F9 B4  B1041.2',),
 ('F9 B4  B1039.2',),
 ('F9 B4 B1045.1',),
 ('F9 B5  B1046.1',),
 ('F9 B4  B1043.2',),
 ('F9 B4  B1040.2',),
 ('F9 B4 B1045.2',),
 ('F9 B5B1047.1',),
 ('F9 B5B1048.1',),
 ('F9 B5 B1046.2',),
 ('F9 B5B1049.1',),
 ('F9 B5 B1048.2',),
 ('F9 B5 B1047.2',),
 ('F9 B5 B1046.3',),
 ('F9 B5B1050',),
 ('F9 B5B1054',),
 ('F9 B5 B1049.2',),
 ('F9 B5 B1048.3',),
 ('F9 B5B1051.1',),
 ('F9 B5B1056.1 ',),
 ('F9 B5 B1049.3',),
 ('F9 B5 B1051.2 ',),
 ('F9 B5 B1056.2 ',),
 ('F9 B5 B1047.3 ',),
 ('F9 B5 B1048.4',),
 ('F9 B5B1059.1',),
 ('F9 B5 B1056.3 ',),
 ('F9 B5 B1049.4',),
 ('F9 B5 B1046.4',),
 ('F9 B5 B1051.3',),
 ('F9 B5 B1056.4',),
 ('F9 B5 B1059.2',),
 ('F9 B5 B1048.5',),
 ('F9 B5 B1051.4',),
 ('F9 B5B1058.1 ',),
 ('F9 B5 B1049.5',),
 ('F9 B5 B1059.3',),
 ('F9 B5B1060.1',),
 ('F9 B5 B1058.2 ',),
 ('F9 B5 B1051.5',),
 ('F9 B5 B1049.6',),
 ('F9 B5 B1059.4',),
 ('F9 B5 B1060.2 ',),
 ('F9 B5 B1058.3 ',),
 ('F9 B5 B1051.6',),
 ('F9 B5 B1060.3',),
 ('F9 B5B1062.1',),
 ('F9 B5B1061.1 ',),
 ('F9 B5B1063.1',),
 ('F9 B5 B1049.7 ',),
 ('F9 B5 B1058.4 ',)]

### Task 9


##### List the records which will display the month names, failure landing_outcomes in drone ship ,booster versions, launch_site for the months in year 2015.

**Note: SQLLite does not support monthnames. So you need to use  substr(Date, 6,2) as month to get the months and substr(Date,0,5)='2015' for year.**


In [35]:
%sql select substr("Date",6,2) as month,"Landing_Outcome","Booster_Version","Launch_Site" from SPACEXTABLE where  "Landing_Outcome"='Failure (drone ship)' and substr("Date",0,5)='2015';

 * sqlite:///my_data1.db
Done.


[('01', 'Failure (drone ship)', 'F9 v1.1 B1012', 'CCAFS LC-40'),
 ('04', 'Failure (drone ship)', 'F9 v1.1 B1015', 'CCAFS LC-40')]

### Task 10




##### Rank the count of landing outcomes (such as Failure (drone ship) or Success (ground pad)) between the date 2010-06-04 and 2017-03-20, in descending order.


In [38]:
%sql select "Landing_Outcome", count(*) from SPACEXTABLE where date between '2010-06-04' and '2017-03-20' group by "Landing_Outcome" order by count(*) desc;

 * sqlite:///my_data1.db
Done.


[('No attempt', 10),
 ('Success (drone ship)', 5),
 ('Failure (drone ship)', 5),
 ('Success (ground pad)', 3),
 ('Controlled (ocean)', 3),
 ('Uncontrolled (ocean)', 2),
 ('Failure (parachute)', 2),
 ('Precluded (drone ship)', 1)]

### Reference Links

* <a href ="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/Labs_Coursera_V5/labs/Lab%20-%20String%20Patterns%20-%20Sorting%20-%20Grouping/instructional-labs.md.html?origin=www.coursera.org">Hands-on Lab : String Patterns, Sorting and Grouping</a>  

*  <a  href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/Labs_Coursera_V5/labs/Lab%20-%20Built-in%20functions%20/Hands-on_Lab__Built-in_Functions.md.html?origin=www.coursera.org">Hands-on Lab: Built-in functions</a>

*  <a  href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/Labs_Coursera_V5/labs/Lab%20-%20Sub-queries%20and%20Nested%20SELECTs%20/instructional-labs.md.html?origin=www.coursera.org">Hands-on Lab : Sub-queries and Nested SELECT Statements</a>

*   <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/Module%205/DB0201EN-Week3-1-3-SQLmagic.ipynb">Hands-on Tutorial: Accessing Databases with SQL magic</a>

*  <a href= "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/Module%205/DB0201EN-Week3-1-4-Analyzing.ipynb">Hands-on Lab: Analyzing a real World Data Set</a>




## Author(s)

<h4> Lakshmi Holla </h4>


## Other Contributors

<h4> Rav Ahuja </h4>


## Change log
| Date | Version | Changed by | Change Description |
|------|--------|--------|---------|
| 2024-07-10 | 1.1 |Anita Verma | Changed Version|
| 2021-07-09 | 0.2 |Lakshmi Holla | Changes made in magic sql|
| 2021-05-20 | 0.1 |Lakshmi Holla | Created Initial Version |


## <h3 align="center"> © IBM Corporation 2021. All rights reserved. <h3/>
